Add checkPhase for pipenv

zupo opened the pull request:
###### Motivation for this change

This is to prevent regressions such as https://github.com/NixOS/nixpkgs/issues/73254
by using pipenv to install a simple Python package, thus testing
that pipenv was built correctly.

I tested the patch locally against nixpgs master (pipenv built successfully) and against nixpgs 19.09 (pipenv build failed, as expected, with `ModuleNotFoundError: No module named 'pip._internal.main'`). 

Refs https://github.com/NixOS/nixpkgs/issues/73254

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @FRidh

Many thanks to @Infinisil and @domenkozar for hand-holding!

zupo committed:
Add checkPhase for pipenv

This is to prevent regressions such as https://github.com/NixOS/nixpkgs/issues/73254
by using pipenv to install a simple Python package, thus testing
that pipenv was built correctly.

Many thanks to @Infinisil and @domenkozar for hand-holding!

FRidh commented on pkgs/development/tools/pipenv/default.nix:
```diff
@@ -39,7 +39,11 @@ in buildPythonApplication rec {
+  doCheck = true;
+  checkPhase = ''
+    export HOME=$PWD
+    $out/bin/pipenv install ${fetchPypi {pname="pyjokes"; version="0.6.0"; sha256="08860eedb78cbfa4618243c8db088f21c39823ece1fdaf0133e52d9c56e981a5";} }
```
Could you replace the whole `${fetchPypi ...}` part with `${wheel.src}`. For that one we already have all the meta set up.

zupo replied:
With 

```diff
-    $out/bin/pipenv install ${fetchPypi {pname="pyjokes"; version="0.6.0"; sha256="08860eedb78cbfa4618243c8db088f21c39823ece1fdaf0133e52d9c56e981a5";} }
+    $out/bin/pipenv install ${wheel.src}
```

I get the following error:

```bash
running install tests
Creating a virtualenv for this project…
Pipfile: /private/var/folders/ck/dlscb24n4lzdlzpt3w0pdcgh0000gn/T/nix-build-pipenv-2018.11.26.drv-0/pipenv-2018.11.26/Pipfile
Using /nix/store/9a5bl4gvzdm69fvhwbwln5ld6vn55iyx-python3-3.7.5/bin/python3.7 (3.7.5) to create virtualenv…
⠇ Creating virtual environment...Already using interpreter /nix/store/9a5bl4gvzdm69fvhwbwln5ld6vn55iyx-python3-3.7.5/bin/python3.7
Using base prefix '/nix/store/9a5bl4gvzdm69fvhwbwln5ld6vn55iyx-python3-3.7.5'
New python executable in /private/var/folders/ck/dlscb24n4lzdlzpt3w0pdcgh0000gn/T/nix-build-pipenv-2018.11.26.drv-0/pipenv-2018.11.26/.local/share/virtualenvs/pipenv-2018.11.26-kbPUpQ_H/bin/python3.7
Also creating executable in /private/var/folders/ck/dlscb24n4lzdlzpt3w0pdcgh0000gn/T/nix-build-pipenv-2018.11.26.drv-0/pipenv-2018.11.26/.local/share/virtualenvs/pipenv-2018.11.26-kbPUpQ_H/bin/python
Installing setuptools, pip, wheel...
done.
Running virtualenv with interpreter /nix/store/9a5bl4gvzdm69fvhwbwln5ld6vn55iyx-python3-3.7.5/bin/python3.7

✔ Successfully created virtual environment! 
Virtualenv location: /private/var/folders/ck/dlscb24n4lzdlzpt3w0pdcgh0000gn/T/nix-build-pipenv-2018.11.26.drv-0/pipenv-2018.11.26/.local/share/virtualenvs/pipenv-2018.11.26-kbPUpQ_H
Creating a Pipfile for this project…
Installing /nix/store/snai9k7dzscb4qpn7fbizspx7zbdfirv-wheel-0.33.6-source…
error: could not create 'wheel.egg-info': Permission denied
```

FRidh reviewed (changes requested):
Where will this test install to? We do not want it to create any additional content in `$out`.

zupo commented:
> Where will this test install to? We do not want it to create any additional content in $out.

It seems $out does not get polluted. Which, according to my very basic understanding, is expected, otherwise most packages' `checkPhase` tests would potentially pollute their $out.

```
[nix-shell:/nix/store/l8q8pp62v2a87qzzp70ddsl07i6pmzb6-pipenv-2018.11.26]$ grep -ri joke ./
grep: warning: GREP_OPTIONS is deprecated; please use an alias or script

[nix-shell:/nix/store/l8q8pp62v2a87qzzp70ddsl07i6pmzb6-pipenv-2018.11.26]$ ls lib/python3.7/site-packages/
pipenv	pipenv-2018.11.26.dist-info

[nix-shell:/nix/store/l8q8pp62v2a87qzzp70ddsl07i6pmzb6-pipenv-2018.11.26]$ ls bin/
pipenv	pipenv-resolver
```

Infinisil commented on pkgs/development/tools/pipenv/default.nix:
```diff
@@ -39,7 +39,11 @@ in buildPythonApplication rec {
-  doCheck = false;
+  doCheck = true;
+  checkPhase = ''
+    export HOME=$PWD
```
I'd do `export HOME=$(mktemp -d)` instead so the build directory doesn't get polluted

zupo commented:
With 

```diff
-    $out/bin/pipenv install ${fetchPypi {pname="pyjokes"; version="0.6.0"; sha256="08860eedb78cbfa4618243c8db088f21c39823ece1fdaf0133e52d9c56e981a5";} }
+    $out/bin/pipenv install ${wheel.src}
```

I get the following error:

```bash
running install tests
Creating a virtualenv for this project…
Pipfile: /private/var/folders/ck/dlscb24n4lzdlzpt3w0pdcgh0000gn/T/nix-build-pipenv-2018.11.26.drv-0/pipenv-2018.11.26/Pipfile
Using /nix/store/9a5bl4gvzdm69fvhwbwln5ld6vn55iyx-python3-3.7.5/bin/python3.7 (3.7.5) to create virtualenv…
⠇ Creating virtual environment...Already using interpreter /nix/store/9a5bl4gvzdm69fvhwbwln5ld6vn55iyx-python3-3.7.5/bin/python3.7
Using base prefix '/nix/store/9a5bl4gvzdm69fvhwbwln5ld6vn55iyx-python3-3.7.5'
New python executable in /private/var/folders/ck/dlscb24n4lzdlzpt3w0pdcgh0000gn/T/nix-build-pipenv-2018.11.26.drv-0/pipenv-2018.11.26/.local/share/virtualenvs/pipenv-2018.11.26-kbPUpQ_H/bin/python3.7
Also creating executable in /private/var/folders/ck/dlscb24n4lzdlzpt3w0pdcgh0000gn/T/nix-build-pipenv-2018.11.26.drv-0/pipenv-2018.11.26/.local/share/virtualenvs/pipenv-2018.11.26-kbPUpQ_H/bin/python
Installing setuptools, pip, wheel...
done.
Running virtualenv with interpreter /nix/store/9a5bl4gvzdm69fvhwbwln5ld6vn55iyx-python3-3.7.5/bin/python3.7

✔ Successfully created virtual environment! 
Virtualenv location: /private/var/folders/ck/dlscb24n4lzdlzpt3w0pdcgh0000gn/T/nix-build-pipenv-2018.11.26.drv-0/pipenv-2018.11.26/.local/share/virtualenvs/pipenv-2018.11.26-kbPUpQ_H
Creating a Pipfile for this project…
Installing /nix/store/snai9k7dzscb4qpn7fbizspx7zbdfirv-wheel-0.33.6-source…
error: could not create 'wheel.egg-info': Permission denied
```

zupo committed:
make sure not to pollute the build directory

Infinisil commented:
It seems that `pipenv` wants to write to the directory it gets in some cases, but `${wheel.src}` is a path in `/nix/store`, which isn't writable. To get around this:
```
cp -r --no-preserve=mode ${wheel.src} $PWD/wheel-src
$out/bin/pipenv install $PWD/wheel-src
```

zupo committed:
Use wheel instead of pyjokes

We need an egg to install to test that pipenv works. Instead of
downloading & installing pyjokes, let's use wheel since we already
have it.

Refs https://github.com/NixOS/nixpkgs/pull/73349#discussion_r345830233

Infinisil commented:
@GrahamcOfBorg build pipenv

FRidh commented:
ugh should have squashed these commits...
